rfcs/0000-state-separation.md

```diff
@@ -0,0 +1,706 @@
+* `Application` is now responsible for driving a single `Dispatcher` through a locally stored `GameData`.
+
+The second part is a big one. That means there's no need to make use of handler abstractions (e.g. `SimpleState`), which makes the number of abstractions used for implementing handlers fewer.
+
+Here is a list of some of the signatures of `StateHandler`:
+
+```rust
+trait StateHandler<S, E> {
+    fn on_start(&mut self, _: &mut World) {
+    }
+
+    fn update(&mut self, _: &mut World) -> Trans<S> {
+        Trans::None
+    }
+
+    fn handle_event(&mut self, _: &mut World, _: &E) -> Trans<S> {
+        Trans::None
+    }
+
+    // snip
+}
+```
+
+Most notably `Trans` has been simplified to only take the `State`, and we no longer receive a parameterized `StateData<T>`.
+Events are passed by reference, since they are passed to multiple handlers (shadow, global, state).
+
+This also means that we need to _associate_ a state with its corresponding handler.
+This is done when building the application like this:
+
+```rust
+let mut game = Application::build("./assets")?
+    .with_state(State::Main, MainState)?
+    .build(GameDataBuilder::default())?;
+
+game.run();
+```
+
+Note that there are a couple of types which come with `State` implementations:
+* `&'static str`
+* `u8`, `u32`, `u64`
+* `i8`, `i32`, `i64`
+* `()` (the empty tuple)
+
+Note that the last one can be used as a "default state" when you only have one in your game:
+
+```rust
+let mut game = Application::build("./assets")?
+    .with_state((), MainState)?
+    .build(GameDataBuilder::default())?;
+
+game.run();
+```
+
+The initial state is determined by its `Default` implementation. So `()` would be `()`, `&str` would be `""` (the empty string).
+
+It's also trivial to implement a HashMap-based state implementation as long as your state implements `Hash + PartialEq + Eq`:
+
+```rust
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+struct SomeStruct {
+    // snip
+}
+
+impl<E> State<E> for SomeStruct {
+    type Storage = MapStateStorage<Self, E>;
+}
+```
+
+For enum-based states there's an even more efficient alternative, which is covered in the next section.
+
+## The `State` derive
+
+The most natural way to model states is to use an enum with unit variants. Unit variants are variants which have no structure or tuple body, like this:
+
+```rust
+enum State {
+  Loading,
+  Main,
+}
+```
+
+To provide a `State` implementation for this enum you can use the `State` derive:
+
+```rust
+#[derive(State, Debug, Clone)]
+enum State {
+  Loading,
+  Main,
+}
+```
+
+This will provide the `State` implementation, and implement `Default` to return the first element.
+
+When using this derive, the Storage implementation is also specialized to an implementation like this:
+
+```rust
+struct State_Storage<E> {
+  f1: Option<Box<dyn amethyst::StateHandler<State, E>>>,
+  f2: Option<Box<dyn amethyst::StateHandler<State, E>>>,
```

Comment:
Would these be set to `None` when we pop a state from the stack?

Reply:
No. States specifically live for the duration of the application. Unless added through the `States` resource.

In principle they could be removed through the same resource as well.

Adding it removing handlers must be done explicitly.